ext/opcache/jit/zend_jit_ir.c

```diff
@@ -0,0 +1,16475 @@
+		} else {
+			jit_set_Z_TYPE_INFO_ref(jit, jit_ZVAL_ADDR(jit, res_addr),
+				ir_ADD_U32(ir_ZEXT_U32(ref), ir_CONST_U32(IS_FALSE)));
+			ir_END_list(end_inputs);
+		}
+	}
+
+	if (!smart_branch_opcode || exit_addr) {
+		if (end_inputs) {
+			ir_MERGE_list(end_inputs);
+		}
+	} else {
+		_zend_jit_merge_smart_branch_inputs(jit, true_label, false_label, true_inputs, false_inputs);
+	}
+
+	return 1;
+}
+
+/* copy of hidden zend_closure */
+typedef struct _zend_closure {
+	zend_object       std;
+	zend_function     func;
+	zval              this_ptr;
+	zend_class_entry *called_scope;
+	zif_handler       orig_internal_handler;
+} zend_closure;
+
+static int zend_jit_stack_check(zend_jit_ctx *jit, const zend_op *opline, uint32_t used_stack)
+{
+	int32_t exit_point = zend_jit_trace_get_exit_point(opline, ZEND_JIT_EXIT_TO_VM);
+	const void *exit_addr = zend_jit_trace_get_exit_addr(exit_point);
+
+	if (!exit_addr) {
+		return 0;
+	}
+
+	// JIT: if (EG(vm_stack_end) - EG(vm_stack_top) < used_stack)
+	ir_GUARD(
+		ir_UGE(
+			ir_SUB_A(ir_LOAD_A(jit_EG(vm_stack_end)), ir_LOAD_A(jit_EG(vm_stack_top))),
+			ir_CONST_ADDR(used_stack)),
+		ir_CONST_ADDR(exit_addr));
+
+	return 1;
+}
+
+static int zend_jit_free_trampoline(zend_jit_ctx *jit, int8_t func_reg)
+{
+	// JIT: if (UNEXPECTED(func->common.fn_flags & ZEND_ACC_CALL_VIA_TRAMPOLINE))
+	ir_ref func = ir_RLOAD_A(func_reg);
+	ir_ref if_trampoline = ir_IF(ir_AND_U32(
+		ir_LOAD_U32(ir_ADD_OFFSET(func, offsetof(zend_function, common.fn_flags))),
+		ir_CONST_U32(ZEND_ACC_CALL_VIA_TRAMPOLINE)));
+
+	ir_IF_TRUE(if_trampoline);
+	ir_CALL_1(IR_VOID, ir_CONST_FC_FUNC(zend_jit_free_trampoline_helper), func);
+	ir_MERGE_WITH_EMPTY_FALSE(if_trampoline);
+
+	return 1;
+}
+
+static int zend_jit_push_call_frame(zend_jit_ctx *jit, const zend_op *opline, const zend_op_array *op_array, zend_function *func, bool is_closure, bool delayed_fetch_this, int checked_stack, ir_ref func_ref, ir_ref this_ref)
+{
+	uint32_t used_stack;
+	ir_ref used_stack_ref = IR_UNUSED;
+	bool stack_check = 1;
+	ir_ref rx, ref, top, if_enough_stack, cold_path = IR_UNUSED;
+
+	ZEND_ASSERT(func_ref != IR_NULL);
+	if (func) {
+		used_stack = zend_vm_calc_used_stack(opline->extended_value, func);
+		if ((int)used_stack <= checked_stack) {
+			stack_check = 0;
+		}
+		used_stack_ref = ir_CONST_ADDR(used_stack);
+	} else {
+		ir_ref num_args_ref;
+		ir_ref if_internal_func = IR_UNUSED;
+
+		used_stack = (ZEND_CALL_FRAME_SLOT + opline->extended_value + ZEND_OBSERVER_ENABLED) * sizeof(zval);
+		used_stack_ref = ir_CONST_ADDR(used_stack);
+
+		if (!is_closure) {
+			used_stack_ref = ir_HARD_COPY_A(used_stack_ref); /* load constant once */
+
+			// JIT: if (EXPECTED(ZEND_USER_CODE(func->type))) {
+			ir_ref tmp = ir_LOAD_U8(ir_ADD_OFFSET(func_ref, offsetof(zend_function, type)));
+			if_internal_func = ir_IF(ir_AND_U8(tmp, ir_CONST_U8(1)));
+			ir_IF_FALSE(if_internal_func);
+		}
+
+		// JIT: used_stack += (func->op_array.last_var + func->op_array.T - MIN(
```

Comment:
I see this is the same code as in `zend_jit_x86.dasc` and `zend_jit_arm64.dasc`, but it looks weird that you assign may_throw to 1 unconditionally, and then have the large conditional below that sets may_throw to 1.

Reply:
> I see this is the same code as in `zend_jit_x86.dasc` and `zend_jit_arm64.dasc`, but it looks weird that you assign may_throw to 1 unconditionally, and then have the large conditional below that sets may_throw to 1.

It looks like the first ``may_throw = 1;`` was originally added by mistake.